src/data_source/fetching.rs

```diff
@@ -1232,22 +1247,31 @@ where
                 };
                 prev_height = block_height;
 
-                // Iterate over all blocks that we should have. Merely iterating over the `Fetch`es
-                // without awaiting them is enough to trigger active fetches of missing blocks,
-                // since we always trigger an active fetch when fetching by block number. Moreover,
-                // fetching the block is enough to trigger an active fetch of the corresponding leaf
-                // if it too is missing.
-                //
-                // The chunking behavior of `get_range` automatically ensures that, no matter how
-                // big the range is, we will release the read lock on storage every `chunk_size`
-                // items, so we don't starve out would-be writers.
+                // Iterate over all blocks that we should have. Fetching the block is enough to
+                // trigger an active fetch of the corresponding leaf if it too is missing. The
+                // chunking behavior of `get_range` automatically ensures that, no matter how big
+                // the range is, we will release the read lock on storage every `chunk_size` items,
+                // so we don't starve out would-be writers.
                 let mut blocks = self
                     .clone()
                     .get_range_with_chunk_size::<_, BlockQueryData<Types>>(
                         chunk_size,
                         start..block_height,
                     );
-                while blocks.next().await.is_some() {}
+                while let Some(fetch) = blocks.next().await {
+                    if let Fetch::Pending(fut) = fetch {
+                        // Wait for the block to be fetched. This slows down the scanner so that we
+                        // don't waste memory generating more active fetch tasks then we can handle
+                        // at a given time. Note that even with this await, all blocks within a
+                        // chunk are fetched in paralle, so this does not block the next block in
+                        // the chunk, only the next chunk until the current chunk completes.
+                        fut.await;
+
+                        // Add a bit of artifical latency to easy pressure on the catchup provider,
+                        // otherwise proactive scans can be highly intensive.
+                        sleep(active_fetch_delay).await;
```

Comment:
@imabdulbasit I'm now thinking we should move this sleep (and the one below on line 1288) into the stream itself. E.g. after we [flatten](https://github.com/EspressoSystems/hotshot-query-service/blob/05743e5e5a09d456aced23e8d9a97b970e448a55/src/data_source/fetching.rs#L1007) the stream of chunks into a stream of individual objets, we could then do something like `.then(|obj| async move { sleep(delay).await; obj })` to inject a small delay between each object. This would have the same effect as these sleeps in slowing down proactive fetching, since we are reading from those streams. But it would also eliminate the issue where subscribing to an old stream leads to very high CPU usage because we scan all the old blocks as fast as the server is able. 

In effect, these two things are the same issue: consuming a stream happens too fast.

OTOH, in the more common usage, where your stream has reached the end of the chain and most of your time is blocked waiting for the next block to arrive, you're not going to notice a small delay anyways. The delay just has to be significantly lower than the expected block time so you can easily keep up with the chain. I think 100ms or less is probably good.

Feel free to push to this branch today if you agree.

Reply:
Or perhaps a delay after each *active fetch* only, as we are doing here, but moved inside the stream itself. Then another delay unconditionally after each chunk. The former limits the requests we make for peers but doesn't slow us down unnecessarily when we already have the data. The latter slows us down a little bit when we are fetching a ton of objects from our own database. Both become negligible when we are passively waiting for new objects to arrive.

Reply:
`But it would also eliminate the issue where subscribing to an old stream leads to very high CPU usage because we scan all the old blocks as fast as the server is able.`

If I am understanding correctly then you mean that we are loading range of blocks from the database constantly and there is no delay between it? If yes then isn't it okay to use max CPU? Maybe we can reduce the chunk size then

Reply:
It can also benefit tho as we don't make concurrent database requests so a delay would provide other requests to be served.

Reply:
Yeah in that case it's technically not a problem to use max CPU...since we're yielding every chunk if we are able to use max CPU it just means everyone else gets to run. It just looks a little scary in Datadog lol

Reply:
1cdd9b77a203b391871f1923264a1dd285771287
Lmk if these changes look good to you. I want to merge this PR and create a new tag before weekly release merge

Reply:
The gist of the changes looks good, but I left a couple of suggestions. Somehow the comments went on the commit page and not the PR, you can see them [here](https://github.com/EspressoSystems/hotshot-query-service/commit/1cdd9b77a203b391871f1923264a1dd285771287